### PR TITLE
Prevent SetDoNotCacheHeaders from overwriting valid cache headers

### PR DESCRIPTION
**Pull request by mikelapierre**
As discussed in issue #3058, this fix prevents SetDoNotCacheHeaders from overwriting the Cache-Control header if no-cache and no-store are already present.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=28679) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=28679) before we can accept your contribution.<br/><br/>:x: mikelapierre [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=28679)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=28679) it.</sub>

File: src/Antiforgery/src/Internal/DefaultAntiforgery.cs [hunk 1]
```diff
@@ -376,39 +376,40 @@ private IAntiforgeryFeature GetTokensInternal(HttpContext httpContext)
...
+            }
 
-            var pragmaHeader = response.Headers[HeaderNames.Pragma];
-            if (!logWarning
-                && !string.IsNullOrEmpty(pragmaHeader)
-                && !string.Equals(pragmaHeader, "no-cache", StringComparison.OrdinalIgnoreCase))
+            var pragmaHeader = httpContext.Response.Headers[HeaderNames.Pragma];
+            if (!logWarning &&
+                !string.IsNullOrEmpty(pragmaHeader) &&
+                !string.Equals(pragmaHeader, "no-cache", StringComparison.OrdinalIgnoreCase))
```
Comment (pranavkm) [inline src/Antiforgery/src/Internal/DefaultAntiforgery.cs:399]: You could avoid an extra lookup if we know we're already logging a warning:

```suggestion
            if (!logWarning &&
                response.TryGetValue(HeaderNames.Pragma, out var pragma) &&
                pragma.Count > 0 &&
                !string.Equals(pragma[0], "no-cache", StringComparison.OrdinalIgnoreCase))
```

File: src/Antiforgery/test/DefaultAntiforgeryTest.cs [hunk 2]
```diff
@@ -348,6 +348,41 @@ public void GetAndStoreTokens_ExistingCachingHeaders_Overriden()
             Assert.Equal("no-cache", context.HttpContext.Response.Headers[HeaderNames.Pragma]);
         }
 
+        [Theory]
+        [InlineData("Cache-Control", "no-cache, no-store, must-revalidate, private")]
+        [InlineData("Pragma", "no-cache")]
+        public void GetAndStoreTokens_ExistingNoCachingHeaders_NotOverriden(string headerName, string headerValue)
```
Comment (pranavkm) [inline src/Antiforgery/test/DefaultAntiforgeryTest.cs:354]: Could you split this as follows: 


* `GetAndStoreTokens_DoesNotOverwriteCacheControlHeader`
* `GetAndStoreTokens_OverwritesCacheControlHeader_IfNoStoreIsNotSet`
* `GetAndStoreTokens_OverwritesCacheControlHeader_IfNoCacheIsNotSet`
* `GetAndStoreTokens_DoesNotOverwritePragmaHeader`

You could use a private method to share all of the boiler plate in `Arrange & Act`.

**Review by pranavkm — COMMENTED** [review]
Thanks, this looks great!

**Review by pranavkm — APPROVED** [review]

**Pull request merged**